README.md
[![GitHub license](https://img.shields.io/github/license/VamsiMudaliar/File-Based-Key-Valued-DataStore-?style=for-the-badge)](https://github.com/VamsiMudaliar/File-Based-Key-Valued-DataStore-/blob/master/LICENSE.txt)
[![GitHub stars](https://img.shields.io/github/stars/VamsiMudaliar/File-Based-Key-Valued-DataStore-?style=for-the-badge)](https://github.com/VamsiMudaliar/File-Based-Key-Valued-DataStore-/stargazers)
[![GitHub issues](https://img.shields.io/github/issues/VamsiMudaliar/File-Based-Key-Valued-DataStore-?style=for-the-badge)](https://github.com/VamsiMudaliar/File-Based-Key-Valued-DataStore-/issues)
[![GitHub forks](https://img.shields.io/github/forks/VamsiMudaliar/File-Based-Key-Valued-DataStore-?style=for-the-badge)](https://github.com/VamsiMudaliar/File-Based-Key-Valued-DataStore-/network) 

[![forthebadge made-with-python](http://ForTheBadge.com/images/badges/made-with-python.svg)](https://www.python.org/)

__ABOUT__

__Built a file-Based Key-Value DataStore that supports the basic CRD(Create,read and delete) operations with additional Update operation . 
This data store is meant to be used as a local storage for one single process on one laptop. The data store is exposed as a library
to the clients that can instantiate a class and work with the data store.__

The data store will support the following functional requirements

1. It can be initialized using an optional file path. if one is not provided, it will reliably create itself in a resonable location on the laptop

2. A New Key Value pair can be added to the data store using the create operation. The key is always a string- capped at 32 chars. the value is always a JSON object - capped at 16KB

3. If Create is invoked from an existing key, an appropriate error must be returned.

4. A Read operation on a key can be performed by providing a key, and recieving the value in response as a JSON object 

5. A Delete operation can be performed by providing a key

6. Every Key supports setting a TIME TO LIVE property when it is created. This property is optional. if provided it will be evaluated as an integer defining the number of seconds 
   the key must be retained in the data store Once the TIME TO LIVE for a key has epired, the key will no longer be available for read or delete operation

7. Appropriate error messages must always be returned to a client if it uses the data store in unexpected ways or breached any limits


The data Store will also support the following non functional requirements 

1. The size of the file storing the data must never exceed 1GB

2. More than on Client process cannot be allowed to use the same file as a data store at any given time.

3. A client process is allowed to accedd the data store using multiple threads (if it desires to )
   the data must therefore be THREAD-SAFE

4. The client will bear as little memory costs as possible to use this data store while deriving maximum performance with respect to response times for accessing the data store

__To Install run this command on CMD ( on windows )__

```
   pip install LocalDataStore
```

__All the CRUD operations are done on a JSON file which is created automatically when object is created__

> (Optional) Before using this package, try running the test_LocalDataStore file which contains few unit testcase it must roughly take 45 seconds to excute this file. 


**The Package LocalDataStore contains a Class DataStorageDB which contains the main functionality to use it in your code make this import statement**

```
from LocalDataStore import DataStorageDB
```
This Class mainly contains 4 functions which allows you to perform CRUD operations all these methods return a string stating the status of the operation

```

object = DataStorageDB()


Note: that the last parameter is optional and KEY should contain alphanumeric characters and cannot be more than 32 CHAR and VALUE can be at max 16 KB size else it reports ERROR

object.create("KEY","VALUE","TIME_TO_LIVE (in seconds) ")

desc: makes a {key,[value,TTL]} pair if the key is not present already else it reports ERROR

object.delete("KEY") 

desc: deletes the record with given key reports ERROR in case when there is no KEY present in DB or if the TLL property EXPIRES

object.read("KEY")

desc : reads the value of respective KEY provided the TTL still exists else it reports ERROR


object.update("KEY","VALUE","TIME_TO_LIVE (in seconds)")

desc : updates the KEY with the VALUE provided if in DB or if the TTL didnt expire in all other cases it throws error

```


CHANGELOG.txt
Change Log 
=============

1.5.1 (31/12/2020)
---------------------------

- First Release


test_LocalDataStore.py
import unittest
import os
import time
from LocalDataStore import LocalStorageDB



app = LocalStorageDB()

class TestLocalDataStore(unittest.TestCase):
	
	def setUp(self):
		self.case_create_1 = app.create("Vamsi",23,60)
		self.case_create_2 = app.create("Rahul",32,10)
		self.case_create_3 = app.create("Mike",43,55)
		self.case_create_4 = app.create("Vamsi21",23,34)
		self.case_create_5 = app.create("Tokyo",22)

		self.case_delete_1 = app.delete("Vamsi")
		
		#self.case_read_1 = app.read("Rahul")
		time.sleep(11)

		self.case_delete_2 = app.delete("Rahul")
		
		self.case_read_1 = app.read("Tokyo")

		self.case_read_2 = app.read("Someone")

		self.case_update_1 = app.update("Tokyo",120,234)

		self.case_update_2 = app.update("Someone_not in list",20)

	def tearDown(self):
		pass

	def test_create(self):
		self.assertEqual(self.case_create_1,'Key Inserted Successfully.')

		self.assertEqual(self.case_create_2,'Key Inserted Successfully.')

		self.assertEqual(self.case_create_3,'Key Inserted Successfully.')

		self.assertEqual(self.case_create_4,'ERROR : Key Name not Valid ! \n Note : Key Name should only contain alphabets')

		self.assertEqual(self.case_create_5,'Key Inserted Successfully.')



	def test_delete(self):
		self.assertEqual(self.case_delete_1,'Key Deleted... \n')

		self.assertEqual(self.case_delete_2,"ERROR : Couldn't Delete, Key timer has Already Expired ... \n")#'ERROR : Couldn\'t Delete, Key timer has Already Expired ... \n')
		

	def test_read(self):
		#self.assertEqual(self.case_read_1,"ERROR : Timer Expired Key No Longer Present in Database")
		
		#self.assertEqual(self.case_read_2,"ERROR : Key Not Found in Database ! ")
		pass

	def test_update(self):
		#self.assertEqual(self.case_update_1,"ERROR : Couldn't Update, Key has Already Expired \n")

		#self.assertEqual(self.case_update_2,"ERROR : Key Not Found in Database")
		pass


if __name__ =='__main__':
	unittest.main()

LocalDataStore/main_logic.py
import time
from threading import Thread
import json
from json.decoder import JSONDecodeError

FILE_LIMIT = 1024*1024*1024 # 1GB

KEY_LIMIT = 32 # max 32 characters

VALUE_LIMIT = 16*1024*1024 #16KB

MAX_CONNECTIONS = 1 # no of connections

DEFAULT_LOCATION = "./"

cache={} # our local DB
	

PATH_FORMED = DEFAULT_LOCATION + "new_obj.json"

class LocalStorageDB():

	def __init__(self):
		pass

	def create(self,key,value,time_to_live=0):
		
				# load on the data from a json file
		try:
			with open(PATH_FORMED) as f:
						
				old_cache=json.load(f)

				new_cache=old_cache

				if key in new_cache:				
					print(f"ERROR : Key {key} Already Present...")

				elif key.isalpha():

					curr_record=[]

					if len(cache)<FILE_LIMIT and value<=VALUE_LIMIT:

						if time_to_live==0: 
									#set to default
							curr_record=[value,time_to_live]
						else:
									
							curr_record=[value,time.time()+time_to_live]
									# check if key is atmost 32 chars
									
							if len(key)<=KEY_LIMIT:
									
								new_cache[key]=curr_record

								with open(PATH_FORMED,"w") as fl:
									json.dump(new_cache,fl,indent=2)

								print(f"Key {key} Inserted Successfully.")
							else:
								print("ERROR : Memory Limit Exceed !  \n Note : File size and inserting value should be no more than 1GB and 16KB ")
					else:
						print(f"ERROR : Key Name not Valid ! \n Note : Key Name should only contain alphabets and can be at max {KEY_LIMIT} size")
							#print(f"Note : Key Name should only contain alphabets and can be at max {KEY_LIMIT} size")
		except JSONDecodeError as er:
			pass
	# READ Operation in the file

	def read(self,key):
		
		
			try:

				with open(PATH_FORMED) as f:
					
					old_cache=json.load(f)
					
					new_cache=old_cache

					if key not in new_cache:
						print("ERROR : Key Not Found in Database ! ")

					else:
						store_tmp = new_cache[key]

						pattern = "{ " + str(key)+": "+str(store_tmp[0]) + " }"

						if store_tmp[1]!=0:
								
							if time.time() < store_tmp[1]: 
								return pattern
							else:
								print("ERROR : Timer Expired Key No Longer Present in Database")
						else:
							return pattern
			except JSONDecodeError as er:
				pass



	def delete(self,key):
		
		try:

			with open(PATH_FORMED) as f:
				old_cache=json.load(f)

				new_cache=old_cache

				if key not in new_cache:
					print(f"ERROR : Key {key} Not Found in Database \n")
				
				else:
					store_tmp = new_cache[key]

					if store_tmp[1]!=0:
						if time.time()<store_tmp[1]:
							del new_cache[key]
							print(f"Key {key} Deleted... \n")
						else:
							print(f"ERROR : Couldn't Delete, Key {key} has Already Expired ... \n")
					else:
						del new_cache[key]
						print(f"Key {key} Deleted... \n")

					with open(PATH_FORMED,'w') as f:
						json.dump(new_cache,f,indent=2)

		except JSONDecodeError as er:
			pass

	# Update Function to Update the key Value before it expires

	def update(self,key , value, timer=0):
		
		try:
			with open(PATH_FORMED) as ft:
				old_cache = json.load(ft)

				new_cache=old_cache

				if key not in new_cache:
					print("ERROR : Key Not Found in Database")

				else:
					store_tmp = new_cache[key]

					if time.time()<store_tmp[1] or store_tmp[1]==0:
						
						if timer>0:
							new_value = [value,timer]
						else:
							new_value = [value,store_tmp[1]]

						new_cache[key] = new_value

						with open(PATH_FORMED,"w") as fte:
							json.dump(new_cache,fte,indent=2)

						print("Values Updated..")
					else:
						print("ERROR : Couldn't Update, Key has Already Expired \n")

		except JSONDecodeError as er:
			pass



LocalDataStore/__init__.py
import time
from threading import Thread
import json
from json.decoder import JSONDecodeError
import os


FILE_LIMIT = 1024*1024*1024 # 1GB

KEY_LIMIT = 32 # max 32 characters

VALUE_LIMIT = 16*1024*1024 #16KB

MAX_CONNECTIONS = 1 # no of connections

DEFAULT_LOCATION = os.getcwd() + "\\"

cache={} # our local DB

PATH_FORMED = DEFAULT_LOCATION + "new_obj.json"


class LocalStorageDB():

	def __init__(self):
		if not os.path.exists(PATH_FORMED):
			with open(PATH_FORMED,"w") as fp:
				json.dump(cache,fp,indent=2)

	def create(self,key,value,time_to_live=0):
		
				# load on the data from a json file
		try:
			with open(PATH_FORMED) as f:
						
				old_cache=json.load(f)

				new_cache=old_cache

				if key in new_cache:				
					return "ERROR : Key Already Present..."

				elif key.isalpha():

					curr_record=[]

					if len(cache)<FILE_LIMIT and value<=VALUE_LIMIT:

						if time_to_live==0: 
									#set to default
							curr_record=[value,time_to_live]
						else:
									
							curr_record=[value,time.time()+time_to_live]
									# check if key is atmost 32 chars
									
						if len(key)<=KEY_LIMIT:
									
							new_cache[key]=curr_record

							with open(PATH_FORMED,"w") as fl:
								json.dump(new_cache,fl,indent=2)
							
							return "Key Inserted Successfully."
							
						else:
							return "ERROR : Key Size Exceeds Limit (32)"
					else:
							return "ERROR : Memory Limit Exceed !  \n Note : File size and inserting value should be no more than 1GB and 16KB "
				else:
					return "ERROR : Key Name not Valid ! \n Note : Key Name should only contain alphabets"
							#print(f"Note : Key Name should only contain alphabets and can be at max {KEY_LIMIT} size")
		except JSONDecodeError as er:
			pass
	# READ Operation in the file

	def read(self,key):
		
		
			try:

				with open(PATH_FORMED) as f:
					
					old_cache=json.load(f)
					
					new_cache=old_cache

					if key not in new_cache:
						return "ERROR : Key Not Found in Database ! "

					else:
						store_tmp = new_cache[key]

						pattern = "{ " + str(key)+": "+str(store_tmp[0]) + " }"

						if store_tmp[1]!=0:
								
							if time.time() < store_tmp[1]: 
								return pattern
							else:
								return "ERROR : Timer Expired Key No Longer Present in Database"
						else:
							return pattern
			except JSONDecodeError as er:
				pass



	def delete(self,key):
		
		try:

			with open(PATH_FORMED) as f:
				old_cache=json.load(f)

				new_cache=old_cache

				if key not in new_cache:
					return "ERROR : Key Not Found in Database \n"
				
				else:
					store_tmp = new_cache[key]

					if store_tmp[1]!=0:
						if time.time()<store_tmp[1]:
							del new_cache[key]
							return "Key Deleted... \n"
						else:
							return "ERROR : Couldn't Delete, Key timer has Already Expired ... \n"
					else:
						del new_cache[key]
						return "Key Deleted... \n"

					with open(PATH_FORMED,'w') as f:
						json.dump(new_cache,f,indent=2)

		except JSONDecodeError as er:
			pass

	# Update Function to Update the key Value before it expires

	def update(self,key , value, timer=0):
		
		try:
			with open(PATH_FORMED) as ft:
				old_cache = json.load(ft)

				new_cache=old_cache

				if key not in new_cache:
					return "ERROR : Key Not Found in Database"

				else:
					store_tmp = new_cache[key]

					if time.time()<store_tmp[1] or store_tmp[1]==0:
						
						if timer>0:
							new_value = [value,timer]
						else:
							new_value = [value,store_tmp[1]]

						new_cache[key] = new_value

						with open(PATH_FORMED,"w") as fte:
							json.dump(new_cache,fte,indent=2)

						return "Values Updated.."
					else:
						return "ERROR : Couldn't Update, Key has Already Expired \n"

		except JSONDecodeError as er:
			pass

